### ISSUE TITLE
MAP TX2 Textures

### ISSUE THREAD

**Issue by TheStolenBattenberg** [TheStolenBattenberg/ToolsForKFIV#5]
The TX2 texture sheet in .map files is different from regular TX2 files. Bits per pixel byte says it should be high colour (meaning argb1555) but the data size is of a 32 bit true colour image (data size calculated with width * height * 4).

When the data is extracted and converted to PNG, it's a random mess of colour with some vague block pattern visible. Sometimes something that actually looks like an image of some sorts is visible, in shape only. I don't think the data is compressed as explained in the above paragraph, but maybe something weird is going on here.

In PS2 hardware manuals you can find a section (will cite later) where it describes loading texture pages directly into memory, which I think these might be, as those texture pages are stored as 32 Bit colour regardless of the format... Though using TiledGGD with both 8 & 4 bit image modes enabled and a gradient palette didn't show anything.

**Comment by TheStolenBattenberg** [comment]
Part of the PS2 hardware manuals I talked about above:

> Transfer Textures in 32-bit Mode for Best Speed 
Textures can be transferred in a number of different formats, but the fastest is 32-bit mode. It is possible to transfer textures in this mode even if they are not 32-bit. Such textures can be 'pre-swizzled' into the correct format for 8 and 4bit modes prior to transfer to the GS. This allows you to transfer them in 32bit mode and obtain high transfer speed to the GS. For 8-bit and 4-bit transfer modes the transfer is slowed due to the GS having to convert format internally.
From Graphics Synthesizer Starting Guide.pdf, SDK 303

**Comment by TheStolenBattenberg** [comment]
### TL;DR 
An implementation for map textures will be implemented soon.

### Not TL;DR
After a year or so of light research into the data format used for storing the maps textures, it has been fully confirmed that these are a combination of multiple different bpp and indexed/non-indexed images, stored as a 32bpp ARGB texture. This is done to increase speed of loading as described in the PS2 Hardware manuals.

Unfortunately, this wipes out any hope of extracting these textures through regular means.

A new solution has been thought up (Thanks in part to @IvanDSM) which I am detailing here for future PS2 format reversing dudes in case they come across a similar situation.

### The Theory
Textures on the PS2 can compose of all regular BPP formats, including indexed/non-indexed. They also have a nice feature  that's normally contained in the file formats, which is a 8-byte register containing a location for the pixel data/clut data to be put in PS2 vram.

Primitive packets on the PS2 also have a similar 8-byte register which contains a pixel format, a texture width/height and  a location for the pixel/clut data to be read from PS2 vram. See where this is going?

Basically - If we load all the map textures into a buffer (of identical size to the PS2 vram), then process each primitive packet in each map-piece, grab the 8-byte register and put it into a list if it is not identical to a previously found register.

Then we go through the list we've just populated and we can create a window into our fake PS2 vram buffer using these registers extracted from each piece, using the w/h/bpp/clut loc/pixel loc information provided... After this, it should be as simple as copying the data we've 'windowed' from the big buffer into some smaller buffers, running the conversion functions if they are required... and finally extracting the a regular 32BPP png image.

There are unfortunately some downsides to this method - We miss out on any unused textures that might be hiding out in the files... This is extremely unfortunate as the previous KF games have all had unused textures, and KFIV does seem to have a fair amount of unused content.

It will take a lot longer to convert the map textures too, due to having to essentially emulate part of the PS2 architecture to actually extract them (these 'registers' are what I think are essentially giftags unless my documentation reading has got me confused)

**Comment by TheStolenBattenberg** [comment]
Resolved.
